### PR TITLE
Open directory should focus Finder/Explorer

### PR DESCRIPTION
**Pull request by gabriel**
This will open files into the Finder/Explorer selected or use
openExternal to open a directory in Finder/Explorer.

The openItem call in Electron doesn't grab focus. We now use openExternal.

https://keybase.atlassian.net/browse/DESKTOP-1553

### CODE REVIEW

**Comment by gabriel** [comment]
`shell.openExternal` seemed flakey before because some paths will be symlinks and you need to resolve them before opening, for example, `/keybase/private/gabrielh,chris -> /keybase/private/chris,gabrielh`.

**Comment by gabriel** [comment]
@keybase/react-hackers

File: desktop/app/kbfs-helper.js [hunk 1]
```diff
@@ -1,10 +1,32 @@
...
+      }
+      if (stats.isFile()) {
+        shell.showItemInFolder(path)
+      } else if (stats.isDirectory()) {
+        // Paths if directories might be symlinks.
+        // For example /keybase/private/gabrielh,chris gets redirected to
+        // /keybase/private/chris,gabrielh.
+        fs.realpath(path, (err, resolvedPath) => {
+          if (err) {
+            console.warn('No realpath for %s:', path, err)
```
Comment (chrisnojima) [inline desktop/app/kbfs-helper.js]: I don't think this is the solution yet. Kbfs needs to sort like the gui does else half your folder just won't work which is worse than finder not being in the foreground

Comment (chrisnojima) [inline desktop/app/kbfs-helper.js]: We have a ticket with kbfs to fix their side

Comment (gabriel) [inline desktop/app/kbfs-helper.js]: `openItem` (the current behavior) works with the symlinks, so it's not currently broken (the only issue is that it's not focusing)? Or am I missing something?

File: desktop/app/kbfs-helper.js [hunk 2]
```diff
@@ -1,10 +1,32 @@
...
 
 import {ipcMain, shell} from 'electron'
+import fs from 'fs'
+import {pathToURL} from './paths'
 
 export default function () {
   ipcMain.on('openInKBFS', (e, path) => {
-    // We do the `shell.openItem` here to not block the ui thread
-    shell.openItem(path)
+    fs.stat(path, (err, stats) => {
```
Comment (awendland) [inline desktop/app/kbfs-helper.js]: This hangs my electron instance when checking a path that KBFS hasn't accessed recently. I'm a tad surprised by that because it appears to be a non-blocking call to `fs.stat`, but I've been able to reproduce it several times.

File: desktop/app/kbfs-helper.js (hunk 1 above)
Comment (chrisnojima) [inline desktop/app/kbfs-helper.js]: Yeah. So I think we should leave open item for now. When kbfs changes the canonical path to match what we have in the gui the symlinks problem goes away and we can use openextrernal and it'll work 100% of the time and we'll get focus

File: desktop/app/kbfs-helper.js (hunk 2 above)
Comment (awendland) [inline desktop/app/kbfs-helper.js]: It may actually be a subsequent call that does the hanging, I'm not sure. But one of these `fs` calls hangs the entire app.

Comment (chrisnojima) [inline desktop/app/kbfs-helper.js]: You ca profile the main thread to understand what's happening better. See
electrons docs. We have helpers to do this in our package.json

File: desktop/app/kbfs-helper.js (hunk 1 above)
Comment (strib) [inline desktop/app/kbfs-helper.js]: This fix still seems good to me.  Relying exactly on the TF canonicalization matching exactly between kbfs and the GUI seems brittle in general.  I like this fix -- is there some case where getting the real path won't work?

Comment (gabriel) [inline desktop/app/kbfs-helper.js]: If the user manually types in `gabrielh,chris` instead of `chris,gabrielh`. In that case we need to do realpath. So yeah I think this fix is required in the UI.

File: desktop/app/kbfs-helper.js (hunk 2 above)
Comment (gabriel) [inline desktop/app/kbfs-helper.js]: Is the hang a beach ball, it might be coming from Finder?

File: desktop/app/kbfs-helper.js (hunk 1 above)
Comment (chrisnojima) [inline desktop/app/kbfs-helper.js]: this fix means if the path doesn't match nothing opens up

File: desktop/app/kbfs-helper.js (hunk 2 above)
Comment (awendland) [inline desktop/app/kbfs-helper.js]: Yes it is. I've profiled the app and am finding a pretty large hand there though. Unfortunately, the place it says it's hanging isn't making any sense, so I'm trying to debug more with some console.log/performance.now stuff, but I'm running into some weird issues of not being able to debug the code in devtools.

Comment (awendland) [inline desktop/app/kbfs-helper.js]: The hanging is unrelated to this PR (I think, unless some weird interplay is occurring from the `fs` commands). I've opened up a PR to log what I've seen/tried at https://github.com/keybase/client/pull/3737.

File: desktop/app/kbfs-helper.js (hunk 1 above)
Comment (gabriel) [inline desktop/app/kbfs-helper.js]: Yeah, openExternal will do nothing if the paths don't match.

And when I say manually type, they can do this in the menubar, with "New private folder" or "Open public folder".

So seems like we do need the realpath resolve in the UI?

File: desktop/app/kbfs-helper.js (hunk 2 above)
Comment (gabriel) [inline desktop/app/kbfs-helper.js]: Yeah it looks like KBFS will wait for identify which may have to wait on HTTP requests... I experienced it here, when coinbase took a while to resolve:
https://keybase.atlassian.net/browse/CORE-3488

It seems like the OS beach-balling while keybase is verifying a coinbase proof is not desired behavior?!

File: desktop/app/kbfs-helper.js (hunk 1 above)
Comment (MarcoPolo) [inline desktop/app/kbfs-helper.js]: Does openItem fail too? why the change to openExternal?

Comment (gabriel) [inline desktop/app/kbfs-helper.js]: `openItem` doesn't focus the Finder, so the UX feels a little weird. Also `openItem` uses old AppleScript APIs to workaround some 10.8 bugs, so that's not great especially since we only need 10.10 and above.

Also this change would allow us to open files (that aren't dirs) as selected in the finder, which is a little safer.

Comment (chrisnojima) [inline desktop/app/kbfs-helper.js]: i think we can't use this implementation given the above problem

Comment (chrisnojima) [inline desktop/app/kbfs-helper.js]: its better to have a finder window open not focused than having a majority of your widget rows click and do nothing

**Comment by gabriel** [comment]
OK put a bunch of comments in code explaining the whys.

I guess to summarize:
- We use `openExternal` on directories so that Finder gets focus on macOS
- We use `realpath` to resolve symlinks on directories because openExternal won't work on symlinks.
- We need to do this in the UI, cause openInKBFS can be called from input typed by the user  in the menubar
- We use `showItemInFolder` for files, which opens them in Finder too (and selects them)

If all of the above are true then this PR is OK?

Otherwise unsure how to proceed. Thanks!

**Comment by chrisnojima** [comment]
i do not think we should merge this pr

**Comment by gabriel** [comment]
😢

**Comment by chrisnojima** [comment]
we can't trade one bug for a worse one

**Comment by chrisnojima** [comment]
we will want the file selecting stuff when we udpate the widget to have recent files. but the behavior of doing nothing when the path names don't match is a blocker in my mind. @malgorithms right?

**Comment by gabriel** [comment]
> but the behavior of doing nothing when the path names don't match

But... this PR doesn't break anything. I'm totally confused now.

**Comment by chrisnojima** [comment]
ok i talked to gabriel on hangouts and he says the resolve fixes the names and does allow the finder to open correctly 

👍 🚢  💯
